Add ScanOptions.NativeWindowHandle to scope Scans to rooted UIA sub-tree

blackcatsonly opened the pull request:
#### Details

Adds:
- ScanOptions property settable to IntPtr for HWND of root of subtree to be scanned
- Teach PID descendants search to operate on HWND's subtree instead of GetRootElement
- Add Integration test for test app leaf node scan

##### Motivation

Proposal to address issue #1021 

##### Context

I exposed internals of A11yAutomation to the Integration Tests project so I could find a relevant IUIAutomationElement for testing the ScanOptions' HWND constraint

You could implement this as a "ForWindowHandle()" constraint on the Config, but my external tools use the same scanner repeatedly - it's the individual Scans whose scope needs to be constrained.

#### Pull request checklist
- [X] Addresses an existing issue: #1021

DaveTryon commented on src/Automation/Data/ScanOptions.cs:
```diff
@@ -14,6 +14,11 @@ public class ScanOptions
+        /// <summary>
+        /// The window handle for the root of the UIA subtree to scan.
+        /// </summary>
+        public System.IntPtr WindowHandle { get; set; } = System.IntPtr.Zero;
```
`WindowHandle` seems too vague here. Maybe `ScanRootWindowHandle` would be more descriptive.

blackcatsonly replied:
Agreed.  Updating.

blackcatsonly replied:
Added in latest iteration

DaveTryon commented on src/Automation/Data/ScanOptions.cs:
```diff
@@ -14,6 +14,11 @@ public class ScanOptions
+        /// <summary>
+        /// The window handle for the root of the UIA subtree to scan.
+        /// </summary>
+        public System.IntPtr WindowHandle { get; set; } = System.IntPtr.Zero;
```
`ScanOptions` is intentionally immutable. Rather than adding `ScanTools.SetRootWindowHandle`, maybe pass `ScanTools.ScanRootWindowHandle` as a parameter to `ITargetElementLocator.LocateRootElements` ?

blackcatsonly replied:
Sounds like a reasonable flow to me.  Updating.

blackcatsonly replied:
Added in latest iteration

DaveTryon commented on src/Desktop/UIAutomation/A11yAutomation.cs:
```diff
@@ -147,6 +155,35 @@ public static IEnumerable<DesktopElement> ElementsFromProcessId(int pid, Desktop
             return elements;
         }
 
+        internal static DesktopElement GetFocusedElement()
+        {
+            IUIAutomation uiAutomation = GetDefaultInstance().UIAutomation;
+            IUIAutomationElement focusedElement = uiAutomation.GetFocusedElement();
+            return new DesktopElement(focusedElement, keepElement: true, setMembers: true);
+        }
+
+        internal static DesktopElement GetDepthFirstLastLeafControlElement(DesktopElement rootElement)
+        {
+            var walker = GetDefaultInstance().GetTreeWalker(TreeViewMode.Control);
+            try
+            {
+                IUIAutomationElement leafElement = (IUIAutomationElement)rootElement.PlatformObject;
+                for (IUIAutomationElement currentElement = walker.GetLastChildElement(leafElement); currentElement != null; currentElement = walker.GetLastChildElement(leafElement))
+                {
+                    Marshal.ReleaseComObject(leafElement);
+                    leafElement = currentElement;
+                }
+
+                return leafElement is null
+                    ? rootElement
+                    : new DesktopElement(leafElement, keepElement: true, setMembers: true);
+            }
+            finally
+            {
+                Marshal.ReleaseComObject(walker);
+            }
+        }
+
```
`GetFocusedElement` and `GetDepthFirstLastLeafControlElement` exist purely for the integration tests. As such, I'd argue that they probably should exist in the unit tests

blackcatsonly replied:
Yep, moving this.

blackcatsonly replied:
Added in latest iteration

DaveTryon commented on src/Desktop/UIAutomation/A11yAutomation.cs:
```diff
@@ -14,6 +14,8 @@
 using System.Runtime.InteropServices;
 using UIAutomationClient;
 
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("AutomationTests")]
```
This will fail in the signed build pipeline. Please see https://github.com/microsoft/axe-windows/blob/main/docs/AccessingInternals.md for info on how to make internals visible

blackcatsonly replied:
Thanks, updating.

blackcatsonly replied:
Added in latest iteration

DaveTryon commented on src/Automation/Data/ScanOptions.cs:
```diff
@@ -14,6 +14,11 @@ public class ScanOptions
+        /// <summary>
+        /// The window handle for the root of the UIA subtree to scan.
+        /// </summary>
+        public System.IntPtr WindowHandle { get; set; } = System.IntPtr.Zero;
```
We've tried to expose every `ScanOptions` parameter to the CLI. Is there a similar plan to add that ability for the window handle? It's just a number, so plumbing it through should be straightforward.

blackcatsonly replied:
Added in latest iteration.

codeofdusk commented:
Please update the documentation (`docs/AutomationReference.md` and, for the CLI, the README.md under `src/cli`) with your new options.

blackcatsonly commented on src/Desktop/UIAutomation/A11yAutomation.cs:
```diff
@@ -126,11 +132,13 @@ public static IEnumerable<DesktopElement> ElementsFromProcessId(int pid, Desktop
-                        root = dataContext.A11yAutomation.UIAutomation.GetRootElement();
+                        root = rootWindowHandle == IntPtr.Zero
+                            ? dataContext.A11yAutomation.UIAutomation.GetRootElement()
+                            : dataContext.A11yAutomation.UIAutomation.ElementFromHandle(rootWindowHandle);
```
Good call.  I'll make it fall back to the full process scan.  Adding an integration test.

blackcatsonly replied:
Added in latest iteration

blackcatsonly commented:
@DaveTryon @codeofdusk Thanks for reviewing.  I believe I've addressed all the comments from the first iteration.  Would you please take a second look?

blackcatsonly commented:
@codeofdusk Thanks, missed than one!  Updated in latest iteration.

codeofdusk commented:
/azp run

codeofdusk commented on docs/AutomationReference.md:
```diff
@@ -172,6 +172,7 @@ The `ScanOptions` constructor accepts the following arguments:
 **Name** | **Type** | **Description** | **Default value**
 ---|---|---|---
 scanId | `string` | A string identifier for the scan. If the scan produces output files based on the `Config` object used to create the scanner, the output files will be given the name of the scan id (e.g., MyScanId.a11ytest). | `null`
+scanRootWindowHandle | `IntPtr?` | The native window handle (HWND) of the UIA element whose sub-tree should be scanned. If not specified or no element matches the HWND, the process's full UIA tree will be scanned. | `null`
```
```suggestion
scanRootWindowHandle | `IntPtr` | The native window handle (HWND) of the UIA element whose sub-tree should be scanned. If not specified or no element matches the HWND, the process's full UIA tree will be scanned. | `null`
```

codeofdusk commented on src/Automation/Interfaces/IScanTools.cs:
```diff
@@ -3,6 +3,8 @@
 
 namespace Axe.Windows.Automation
 {
+    using System;
```
Move `using` to outside namespace declaration

codeofdusk commented:
/azp run

codeofdusk commented:
/azp run

codeofdusk commented:
/azp run

blackcatsonly commented:
Thanks @codeofdusk!
